shared/Microsoft.Extensions.ObjectMethodExecutor.Sources/ObjectMethodExecutor.cs

```diff
@@ -0,0 +1,415 @@
+                getResultFunc = Expression.Lambda<Func<object, object>>(
+                    Expression.Constant(null),
+                    getResultParam).Compile();
+            }
+            else
+            {
+                // var getResultFunc = (object awaiter) =>
+                //     (object)((CustomAwaiterType)awaiter).GetResult();
+                getResultFunc = Expression.Lambda<Func<object, object>>(
+                    Expression.Convert(
+                        Expression.Call(
+                            Expression.Convert(getResultParam, awaiterType),
+                            awaiterGetResultMethod),
+                        typeof(object)),
+                    getResultParam).Compile();
+            }
+
+            // var onCompletedFunc = (object awaiter, Action continuation) => {
+            //     ((CustomAwaiterType)awaiter).OnCompleted(continuation);
+            // };
+            var onCompletedParam1 = Expression.Parameter(typeof(object), "awaiter");
+            var onCompletedParam2 = Expression.Parameter(typeof(Action), "continuation");
+            var onCompletedFunc = Expression.Lambda<Action<object, Action>>(
+                Expression.Call(
+                    Expression.Convert(onCompletedParam1, awaiterType),
+                    awaiterOnCompletedMethod,
+                    onCompletedParam2),
+                onCompletedParam1,
+                onCompletedParam2).Compile();
+
+            // return new ObjectMethodExecutorAwaitable(
+            //     (object)getCustomAwaitable(),
+            //     getAwaiterLambda,
+            //     isCompletedLambda,
+            //     getResultLambda,
+            //     onCompletedLambda);
+            var returnValueExpression = Expression.New(
+                _objectMethodExecutorAwaitableConstructor,
+                Expression.Convert(methodCall, typeof(object)),
+                Expression.Constant(getAwaiterFunc),
+                Expression.Constant(isCompletedFunc),
+                Expression.Constant(getResultFunc),
+                Expression.Constant(onCompletedFunc));
+
+            var lambda = Expression.Lambda<ActionExecutorAsync>(returnValueExpression, targetParameter, parametersParameter);
+            return lambda.Compile();
+        }
+        
+        private static object[] GetParameterDefaultValues(ParameterInfo[] parameters)
+        {
+            var values = new object[parameters.Length];
+
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                var parameterInfo = parameters[i];
+                object defaultValue;
+
+                if (parameterInfo.HasDefaultValue)
+                {
+                    defaultValue = parameterInfo.DefaultValue;
+                }
+                else
+                {
+                    var defaultValueAttribute = parameterInfo
+                        .GetCustomAttribute<DefaultValueAttribute>(inherit: false);
```

Comment:
OK, done. I've made supplying defaults optional, since even the ability to store default values seems pretty scenario-specific. So now there are two overloads of `Create` - one that requires `parameterDefaultValues` and one that doesn't.  If you use the one that doesn't, then you can't subsequently call `GetDefaultValueForParameter` (it will throw if you do).